### PR TITLE
Adding additional wrappers for SmartSync

### PR DESCRIPTION
**Pull request by trooper2013**
Still WIP. Tests are pending (next pr).

### CODE REVIEW

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift [hunk 1]
```diff
@@ -32,6 +32,96 @@ enum SmartStoreError : Error {
...
+                                 .build()
+     ```
+     */
+    public class Builder {
+        var queryDict: Dictionary<String,Any> = [[]]
+        
+        public required init() {
+        }
+        
+        public func queryType(value: String) -> Self {
```
Comment (wmathurin) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift]: Shouldn't that take a `SFSoupQueryType`?

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartSyncExtensions.swift [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+         ```
+         - parameters
+             - target: SFSyncDownTarget target
+             - soupName: Soup Name
+             - updateBlock: Block to invoke
+         - Returns: SFSmartStore wrapped in a promise.
+         */
+        func syncDown(target: SFSyncDownTarget, soupName: String, updateBlock: SFSyncSyncManagerUpdateBlock) -> Promise<SFSyncState> {
+            return Promise(.pending) {  resolver in
+                resolver.fulfill(self.api!.syncDown(with: target, soupName: soupName, update: updateBlock))
```
Comment (wmathurin) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartSyncExtensions.swift]: The sync down call returns right away - should the promise resolve when the sync actually completes.
So we could do something similar to what we did in the react native bridge: https://github.com/forcedotcom/SalesforceMobileSDK-iOS/blob/master/libs/SalesforceReact/SalesforceReact/Classes/SFSmartSyncReactBridge.m#L97

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartSyncExtensions.swift [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+         ```
+         - parameters
+         - syncId: NSNumber
+         - updateBlock: Block to invoke
+         - Returns: SFSyncState wrapped in a promise.
+         */
+
+        func reSync(syncId: NSNumber, updateBlock: SFSyncSyncManagerUpdateBlock) -> Promise<SFSyncState> {
+            return Promise(.pending) {  resolver in
+                resolver.fulfill(self.api!.reSync(syncId, update: updateBlock)))
```
Comment (wmathurin) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartSyncExtensions.swift]: The reSync call returns right away - should the promise resolve when the reSync actually completes.
So we could do something similar to what we did in the react native bridge: https://github.com/forcedotcom/SalesforceMobileSDK-iOS/blob/master/libs/SalesforceReact/SalesforceReact/Classes/SFSmartSyncReactBridge.m#L107

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartSyncExtensions.swift [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+import Foundation
+import SmartSync
+import PromiseKit
+
+extension SFSmartSyncSyncManager {
```
Comment (wmathurin) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartSyncExtensions.swift]: What happened to syncUp?

**Comment by brandonpage** [comment]
## Danger has errored
[!] Invalid `Dangerfile-Lib.rb` file: [31mWorkspace file not found at path `/Users/iosadmin/03b72f19/workspace/SalesforceMobileSDK.xcworkspace`[0m
```
 #  from .circleci/Dangerfile-Lib.rb:15
 #  -------------------------------------------
 #  if ENV.has_key?('JENKINS_URL')
 >    xcov.report(
 #        scheme: 'UnitTests',
 #  -------------------------------------------
```

<p align="right" data-meta="generated_by_CI-result">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift [hunk 5]
```diff
@@ -97,6 +248,18 @@ extension SFSmartStore {
             }
         }
         
+        /**
+         Create a soup
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift]: Copy-paste error on doc comment here.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift [hunk 6]
```diff
@@ -161,48 +383,120 @@ extension SFSmartStore {
             }
         }
         
+        /**
+         Lookup the soup entry Id
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift]: Copy-paste error on doc comment here.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift [hunk 7]
```diff
@@ -161,48 +383,120 @@ extension SFSmartStore {
...
+         */
         public func removeEntries(entryIds: [Any], soupName: String) -> Promise<Void> {
             return Promise(.pending) {  resolver in
                 self.api!.removeEntries(entryIds, fromSoup: soupName)
                 resolver.fulfill(())
             }
         }
         
+        /**
+         Lookup the soup entry Id
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift]: Copy-paste error on doc comment here.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift [hunk 8]
```diff
@@ -161,48 +383,120 @@ extension SFSmartStore {
...
+         */
         public func removeSoup(soupName: String) -> Promise<Void>{
             return Promise(.pending) {  resolver in
                 self.api!.removeSoup(soupName)
                 resolver.fulfill(())
             }
         }
         
+        /**
+         Remove a soup
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift]: Should be `Remove all soups`.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift [hunk 9]
```diff
@@ -223,34 +531,94 @@ public class SFSmartStoreClient {
...
+     */
     public class func removeAllSharedStores() -> Promise<Void> {
         return Promise(.pending) { resolver in
             SFSmartStore.removeAllStores()
             resolver.fulfill(())
         }
     }
     
+    /**
+     Remove all shared store instances for current user.
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift]: Copy-paste error on doc comment here. Should be `global`.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartSyncExtensions.swift [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+         - parameter name: Name of sync.
+         */
+        public func deleteSync(name: String) -> Promise<Void>  {
+            return Promise(.pending) {  resolver in
+                resolver.fulfill(self.api!.deleteSync(byName:name))
+            }
+        }
+
+        /**
+         Delete a sync
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartSyncExtensions.swift]: Copy-paste error on doc comment here. Should be `Create sync down`.

**Pull request merged**